### PR TITLE
Fix memory leak in image classification demo, which is caused by dataprovider

### PR DESCRIPTION
**Pull request by qingqing01**
@reyoung We have met this problem before in my experiments. This way can prevent the memory leak if not change the C++ code.  Please feel free to close this issue if you think we should have better solution here. 
Fix issue #310

### CODE REVIEW

**Comment by emailweixu** [comment]
Can you comment why this fix memory leak?

**Comment by reyoung** [comment]
@emailweixu The reason is [here](http://www.paddlepaddle.org/doc_cn/ui/data_provider/pydataprovider2.html#id5)

The issue #310 may have other problems, because the memory should not be increased by pass.  I was trying to reproduce the #310, #311 in my laptop last night, but failed. 

Should we start the `regression test` project, to write automatically scripts to test each demo? Also the `regression test` should record the memory usage, gpu usage, cpu usage, error rate, cost, etc. 

@wen-bo-yang , Could you help us to write `regression test`? We could use `Docker` to run the `regression test`. The sample of code is [here](https://github.com/baidu/Paddle/tree/develop/paddle/scripts/tools/build_docs). This sample script can build documentation inside a docker container, and export the documentation to this directory.

@wen-bo-yang The machine with `docker` installed can be provided. Please contact me if you want help.

File: demo/image_classification/image_provider.py [hunk 1]
```diff
@@ -58,24 +58,26 @@ def hook(settings, img_size, mean_img_size, num_classes, color, meta, use_jpeg,
     settings.logger.info('DataProvider Initialization finished')
 
 
-@provider(init_hook=hook)
-def processData(settings, file_name):
+@provider(init_hook=hook, min_pool_size=0)
```
Comment (reyoung) [inline demo/image_classification/image_provider.py]: `min_pool_size=0` will cause the dataset is not shuffled fully. See [here](https://github.com/baidu/Paddle/blob/develop/doc_cn/faq/index.rst#dataprovider)

File: demo/image_classification/image_provider.py [hunk 2]
```diff
@@ -58,24 +58,26 @@ def hook(settings, img_size, mean_img_size, num_classes, color, meta, use_jpeg,
...
-        else:
-            img = data['images'][i]
-        img_feat = image_util.preprocess_img(img, settings.img_mean,
-                                             settings.img_size, settings.is_train,
-                                             settings.color)
-        label = data['labels'][i]
-        yield img_feat.tolist(), int(label)
+    with open(file_list, 'r') as fdata:
+        for file_name in fdata:
+            data = cPickle.load(io.open(file_name.strip(), 'rb'))
```
Comment (reyoung) [inline demo/image_classification/image_provider.py]: I think that opening a file but not close is a very bad manner.

But when I was trying to find memleak last night, it seems that this part is not the reason.

Maybe we should change these bad manners later, not now.

**Comment by wen-bo-yang** [comment]
@reyoung  OK, I hear your message.  I will design regression test for each demo.

**Comment by reyoung** [comment]
@wen-bo-yang and maybe the image_classification should be the first, because these memleak issues.

**Comment by wen-bo-yang** [comment]
@reyoung OK  I understand.

**Pull request merged**